README.md
DBGP Notepad++ plugin
=====================

Status
------

After a [request from the Notepad++ community][0], this project was created to reboot development of DBGp.
The original plugin is [abandonware][1]. The author now maintains a [VS Code extension][2] with the same and better functionality.

The goals of this project are to:

* make DBGp's code base available in a clean, organized form, while preserving as much of the original as possible
* upgrade the visual components libraries for compatibility with recent Delphi IDEs
* make the code binary compatible with recent 64-bit versions of Notepad++

There are currently __*no*__ plans to implement new features. Users looking for those will be better served by the [VS Code extension][2].

Configuring Xdebug 3
--------------------

By default, plugin versions >= 0.14 will open a TCP connection on port 9003, the same default port as Xdebug 3: https://xdebug.org/docs/upgrade_guide#Step-Debugging

Older versions will listen on port 9000 by default, the same as Xdebug 2.

For any v3 release of Xdebug, a minimal configuration resembles the following:

    [xdebug]
    xdebug.mode = debug
    xdebug.log_level = 10

*Note*
The `xdebug.remote_handler` property is obsolete as of version 2.9: https://2.xdebug.org/docs/all_settings#remote_handler

Xdebug 2 can be configured according to the "SERVER SIDE INSTALL" section of [the original README][3], except the port number must be set explicitly:

    [xdebug]
    xdebug.remote_port = 9003

Alternatively, change the plugin's default port in the settings dialog, or edit the configuration file at `<NPP_PLUGINS_DIRECTORY>\Config\dbgp.ini`:

    [Misc]
    listen_port = 9000

Known Issues
------------

* Xdebug creates malformed URIs for files with Windows UNC paths: <https://bugs.xdebug.org/view.php?id=1964>

  As a result, users may notice that:

  - breakpoints can be set in local files with UNC paths, but they are never hit since the debugger can't find them;
    attempting to remove the breakpoints may also fail

  - sending the `source` command with a UNC path fails with a "cannot open file" error

  As a workaround, map any file shares to a drive letter, e.g. `net use Z: \\hostname\share`, and only send drive-mapped
  file names to the debugger, e.g. `source -f file:///Z:/var/www/index.php`

* Re-positioning the debugger panel too many times may cause an infinite redraw loop, hanging Notepad++.
  This could simply be [a race condition][4], or a limitation of the editor's [window management capabilities][5].
  Independent bug reports found [here][6], [here][7] and [here][8] suggest that it's a longstanding problem, and
  not unique to DBGp. See [5b0af29].

* Tooltips are not always destroyed completely, leaving hollow outlines on the screen. See [6560b64].

[0]: https://community.notepad-plus-plus.org/topic/22772/new-cross-platform-plugin-template-for-delphi-developers/3
[1]: https://github.com/zobo/dbgpPlugin/issues/1#issuecomment-770746125
[2]: https://marketplace.visualstudio.com/items?itemName=xdebug.php-debug
[3]: https://bitbucket.org/rdipardo/dbgp/raw/bf0577b6c621063ff1b82d220afcc32eea9c930e/README.orig.txt
[4]: https://community.notepad-plus-plus.org/topic/13116/creating-a-docked-window-from-a-background-thread
[5]: https://github.com/kbilsted/NotepadPlusPlusPluginPack.Net/issues/17#issuecomment-683459898
[6]: https://sourceforge.net/p/notepad-plus/discussion/482781/thread/ab626469
[7]: https://github.com/kbilsted/NotepadPlusPlusPluginPack.Net/issues/17
[8]: https://community.notepad-plus-plus.org/topic/23667/npp-got-stuck-after-rearranging-dialogs
[5b0af29]: https://bitbucket.org/rdipardo/dbgp/commits/5b0af29
[6560b64]: https://bitbucket.org/rdipardo/dbgp/commits/6560b64


CHANGELOG.txt
2023-12-15 v0.14.2.1

Fixed property evaluation popups in N++ versions 8.4.1/2 by checking for zero-padded version words.
Fixed the tiny size of the toolbar bitmap by dynamically setting its dimensions at runtime.
Moved DLL version info to a resource file.

2023-11-26 v0.14.2

Enabled source mapping of local, drive-mapped files.[^1]
Improved error message reporting.
Added a GUI control to set the debugger's listen port.
The config form now automatically disables 'source' commands when the local setup option is chosen.
The DBGp message input dialog can now save output to a file.

---
[^1]: Files with Windows UNC path names are affected by a known issue: <https://bugs.xdebug.org/view.php?id=1964>

2023-04-20 v0.14.1

Applied a new workaround for the issue where moving docked panels causes an infinite redraw loop.
Tried to prevent more tooltips from leaving visual artifacts on the screen.
Added an icon to the debugger panel's (minimized) window tab.

2022-12-04 v0.14

*** First generally available 64-bit version ***
Changed API calling logic to dynamically use the full-range variants in Notepad++ 8.4.3 and later.
Tried to mitigate an issue where moving a docked panel may cause an infinite redraw loop.
Tried to mitigate an issue where tooltips are not always completely destroyed.

2022-07-24 v0.14 alpha-2

Changed default port to 9003, same as Xdebug 3
Upgraded project to Delphi 10.4
Upgraded JVCL to 3.50
Updated VirtualTreeView to 7.6.2
Updated sockets library to 27.0.40680.4203 (dclsockets270.bpl)

2022-05-01 v0.14 alpha

Added 64-bit platform support.
Updated Scintilla interface to support 64-bit character positions in x64 Notepad++ 8.3 and later.
*** This is NOT backward compatible with older x64 Notepad++ versions ***
Upgraded project to Delphi 10.3.
Updated VirtualTreeView to 7.6.1 from https://github.com/JAM-Software/Virtual-TreeView
Updated sockets library to 26.0.36039.7899 (dclsockets260.bpl)

2012-08-25 v0.13 beta

On demand fetching of lower levels of property tree. If you set max_depth to 1
each new level will be fetched from the debugging engine.
Using AutoScroll=false on forms so that ClientWidth/ClientHeight is used and
forms look better on Windows 7.
Upgraded to JVCL 3.45.
And added fix for http://issuetracker.delphi-jedi.org/view.php?id=5023
Checking to see if dbgp source temp file changed and not rewriting it.

2010-08-12 v0.12 beta

Added max_data to config dialog.
Using NPPN_READY to initialize markers. WM_CREATE is not available in 5.7 anymore.
Updated FuncArray, separators now have Func set to nil.
Mapping logic tries to fall back to direct local mapping if nothing matches.
Small update for NPPM_GETCURRENTWORD.

2009-05-05 v0.11 beta

A quick fix for markers as N++ masks them.
Try to do numerical comparison for sorting keys in Context and other windows.

2008-11-16 v0.10 beta

Changed to UNICODE.
Handling for multiple concurent debug sessions.
More socket error handling.


2008-07-30 v0.9 beta

New configuration, local setup bypasses all mapping.
Experimental breakpoint tracking.
Added GetLongPathName, so that internal paths are aligned with notepad's paths.
Added unescape routine for dbgp files (eval).
Updated breakpoint edit form for better alignment with xdebug.
Remove all breakpoints.
Added copy to clipboard from context and inspect tree forms.
Fixed a bug (or some change in NPP) with opening files (green arrow not shown).
A readme section on testing.

2008-04-20 v0.8 beta

Added listen port config directive. No dialog interface for it, just edit config file.
CTRL+Click on margin to add breakpoint.
Notepad++ will now remember last position of the debug window.
Fixed a problem with newer XDebug when stepping out of the script.
Made variable trees (context, watch...) sortable.
Also when a variable inside an array changes, the red change indicator is propagated back to the top.
Added a stop button. Aborts script.
Watch child now present at startup.

2008-02-16 v0.7 beta fix

Fixed a bug in configuration dialog.

2008-02-08 v0.7 beta

Added config option to not break at first line of code.
A few code cleanups.
Fixed a bug in the watch child.
Unreleased windows path mapping problem.

2007-11-25 v0.6 beta

Added new config options.
Confiuration changes take effect right away.
NPPM_SHUTDOWN.
New breakpoint icons.

2007-10-12 v0.5 alpha fix

Added WM_CLOSE processing that should fix the hanging after exit problem. (Waiting for NPP approval).

2007-10-06 v0.5 alpha

About form displays current version and build.
Breakpoints display icons.
One config variable was miss-named.
Added menu items for all child windows.
Experimental eval on mouse dwell.
Moved plugin config file to where it's supposed to be. Most people will need to reenter their configuration.
Fancy menu enabling and disabling.
Now using file version.
Implemented synced call for easier property getting (watches, mouse dwell).
Small breakpoints bug.
Reuse eval result form.
Some window handling changes.
"Run to" button. (Temporary breakpoints don't work as they should because of XDebug)

src/lib/dbgpnppplugin.pas
{
    This file is part of DBGP Plugin for Notepad++
    Copyright (C) 2007  Damjan Zobo Cvetko

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
}

unit dbgpnppplugin;
{
  This file "extends" the NppPlugin unit and implements
  the startup routines... The main dll handler calls these routines...
}
interface

uses
  NppPlugin, MainForm, nppdockingform,
  ConfigForm, Classes, Windows, Graphics, IniFiles, DbgpWinSocket, AboutForm;

const
  NO_WARN_UNC_BPS = 1;
  NO_WARN_UNC_SRCMAPS = 2;
  MARKER_ARROW = 3;
  MARKER_BREAK = 4;
  MENU_EVAL_INDEX = 8;
type
  TDbgpNppPluginConfig = record
    maps: TMaps;
    refresh_local: boolean;
    refresh_global: boolean;
    use_source: boolean;
    local_setup: boolean;
    start_closed: boolean;
    break_first_line: boolean;
    max_depth: integer;
    max_children: integer;
    listen_port: integer;
    max_data: integer;
  end;
  TDbgpMenuState = ( dmsOff, dmsDisconnected, dmsConnected );
  TDbgpNppPlugin = class(TNppPlugin)
  private
    MainForm: TNppDockingForm1;
    ConfigForm: TConfigForm1;
    AboutForm: TAboutForm1;
    menuEvalIndex: Integer;
    mainFormIcon: TIcon;
    IsCompatible: Boolean;
    procedure GrayFuncItem(i: integer);
    procedure EnableFuncItem(i: integer);
    procedure InitMarkers;
    procedure WarnUser;
  public
    config: TDbgpNppPluginConfig;
    UNCWarnings: 0..3;
    constructor Create;
    destructor Destroy; override;

    procedure BeNotified(sn: PSciNotification); override;

    procedure FuncDebugger;
    procedure FuncConfig;
    procedure FuncStepInto;
    procedure FuncStepOver;
    procedure FuncStepOut;
    procedure FuncRunTo;
    procedure FuncRun;
    procedure FuncStop;
    procedure FuncEval;
    procedure FuncAbout;
    procedure FuncBreakpoint;
    procedure FuncLocalContext;
    procedure FuncGlobalContext;
    procedure FuncStack;
    procedure FuncBreakpoints;
    procedure FuncWatches;
    procedure ReadMaps(var maps: TMaps);
    procedure WriteMaps(conf: TDbgpNppPluginConfig);

    procedure ChangeMenu(state: TDbgpMenuState);
    function Warn(Msg: String; Caption: String = 'Error'; Icon: Cardinal = MB_ICONERROR): Integer;
  end;

var
  Npp: TDbgpNppPlugin;

procedure _FuncDebugger; cdecl;
procedure _FuncConfig; cdecl;
procedure _FuncStepInto; cdecl;
procedure _FuncStepOver; cdecl;
procedure _FuncStepOut; cdecl;
procedure _FuncRunTo; cdecl;
procedure _FuncRun; cdecl;
procedure _FuncStop; cdecl;
procedure _FuncEval; cdecl;
procedure _FuncAbout; cdecl;
procedure _FuncBreakpoint; cdecl;
procedure _FuncLocalContext; cdecl;
procedure _FuncGlobalContext; cdecl;
procedure _FuncStack; cdecl;
procedure _FuncBreakpoints; cdecl;
procedure _FuncWatches; cdecl;

implementation

{ TDbgpNppPlugin }
uses
  SysUtils,Controls;

procedure TDbgpNppPlugin.BeNotified(sn: PSciNotification);
const
  hNil = THandle(Nil);
var
  SciTextRangeMsg,LangId: Cardinal;
  bmpX,icoX,bmpY,icoY: Integer;
  hHDC: HDC;
  tb: TToolbarIcons;
  tbDM: TTbIconsDarkMode;
  tr: TSciTextRange;
  s: string;
  pzS: PAnsiChar;
  i, SLen: Sci_Position;
begin
  if (HWND(sn^.nmhdr.hwndFrom) = self.NppData.NppHandle) then
  begin
    if (sn^.nmhdr.code = NPPN_TBMODIFICATION) then
    begin
      hHDC := hNil;
      try
        hHDC := GetDC(hNil);
        bmpX := MulDiv(16, GetDeviceCaps(hHDC, LOGPIXELSX), 96);
        bmpY := MulDiv(16, GetDeviceCaps(hHDC, LOGPIXELSY), 96);
        icoX := MulDiv(32, GetDeviceCaps(hHDC, LOGPIXELSX), 96);
        icoY := MulDiv(32, GetDeviceCaps(hHDC, LOGPIXELSY), 96);
      finally
        ReleaseDC(hNil, hHDC);
      end;
      tb.ToolbarBmp := LoadImage(Hinstance, 'IDB_DBGP_TEST', IMAGE_BITMAP, bmpX, bmpY, 0);
      tb.ToolbarIcon := LoadImage(Hinstance, 'IDB_DBGP_ICO', IMAGE_ICON, icoX, icoY, (LR_DEFAULTSIZE or LR_LOADTRANSPARENT));
      if SupportsDarkMode then begin
        with tbDM do begin
          ToolbarBmp := tb.ToolbarBmp;
          ToolbarIcon := tb.ToolbarIcon;
          ToolbarIconDarkMode := tb.ToolbarIcon;
        end;
        SendNppMessage(NPPM_ADDTOOLBARICON_FORDARKMODE, self.FuncArray[0].CmdID, @tbDM);
      end else
        SendNppMessage(NPPM_ADDTOOLBARICON_DEPRECATED, self.FuncArray[0].CmdID, @tb);

      self.IsCompatible := {$IFDEF CPUx64}self.SupportsBigFiles{$ELSE}True{$ENDIF};
      if (not self.IsCompatible) then self.WarnUser;
    end;
    if (sn^.nmhdr.code = NPPN_SHUTDOWN) then
    begin
      if (Assigned(self.MainForm)) then self.MainForm.Hide;
    end;
  end;

  if (not self.IsCompatible) then Exit;

  if (sn^.nmhdr.code = NPPN_READY) then
  begin
    self.InitMarkers;
  end;
  if (sn^.nmhdr.code = SCN_DWELLSTART) then
  begin
    //if (Assigned(self.TestForm)) then self.TestForm.OnDwell();
    //ShowMessage('SCN_DWELLSTART '+IntToStr(sn^.position));
    //self.MainForm.state
    SendNppMessage(NPPM_GETCURRENTLANGTYPE, 0, @LangId);
    if not (TNppLang(LangId) in [L_PHP, L_HTML, L_ASP, L_JSP]) then Exit;

    s := '';
    SendMessage(CurrentScintilla, SCI_SETWORDCHARS, 0, LPARAM(PChar('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_$->')));
    tr.chrg.cpMin := SendMessage(CurrentScintilla, SCI_WORDSTARTPOSITION, sn^.position, 0);
    SendMessage(CurrentScintilla, SCI_SETWORDCHARS, 0, LPARAM(PChar('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_')));
    tr.chrg.cpMax := SendMessage(CurrentScintilla, SCI_WORDENDPOSITION, sn^.position, 0);

    if (tr.chrg.cpMin<>-1) and (tr.chrg.cpMax-tr.chrg.cpMin>0) then
    begin
      if Self.HasFullRangeApis then
          SciTextRangeMsg := SCI_GETTEXTRANGEFULL
        else
          SciTextRangeMsg := SCI_GETTEXTRANGE;
      SetLength(s, tr.chrg.cpMax-tr.chrg.cpMin+10);
      tr.lpstrText := PAnsiChar(UTF8Encode(s));
      SLen := SendMessage(CurrentScintilla, SciTextRangeMsg, 0, LPARAM(@tr));
      SetString(s, PAnsiChar(tr.lpstrText), SLen);
      pzS := PAnsiChar(UTF8Encode(s));
      SendMessage(CurrentScintilla, SCI_CALLTIPSHOW, sn^.position, LPARAM(pzS+' = Getting...'));
      SendMessage(CurrentScintilla, SCI_SETCHARSDEFAULT, 0, 0);
      if (s<>'') then
      begin
        s := self.MainForm.sock.GetPropertyAsync(UTF8ToString(pzS));
        SendMessage(CurrentScintilla, SCI_CALLTIPSHOW, sn^.position, LPARAM(UTF8Encode(s)));
      end;
    end;
    if (s = '') then
        SendMessage(CurrentScintilla, SCI_CALLTIPCANCEL, 0, 0);
  end;

  if (sn^.nmhdr.code = SCN_DWELLEND) then
  begin
    //add a delay somehow...
    //SendMessage(self.NppData.ScintillaMainHandle, SCI_CALLTIPCANCEL, 0, 0);
  end;

  //if (sn^.nmhdr.code = SCN_DOUBLECLICK) then ShowMessage('SCN_DOUBLECLICK');
  if (sn^.nmhdr.code = SCN_MARGINCLICK) and (sn^.margin = 1) and (sn^.modifiers and SCMOD_CTRL = SCMOD_CTRL) then
  begin
    if (Assigned(self.MainForm)) then
    begin
      self.GetFileLine(s,i);
      i := SendMessage(CurrentScintilla, SCI_LINEFROMPOSITION, sn.position, 0);
      self.MainForm.ToggleBreakpoint(s,i+1);
      //ShowMessage('SCN_MARGINCLICK '+IntToStr(i));
    end;
  end;

  if (sn^.nmhdr.code = SCN_MODIFIED) and (sn^.modificationType and SC_MOD_CHANGEMARKER = SC_MOD_CHANGEMARKER) then
  begin
//    ShowMessage('SCN_MODIFIED SC_MOD_CHANFEMARKER '+IntToStr(sn^.line));
  end;
end;

constructor TDbgpNppPlugin.Create;
var
  sk: PShortcutKey;
begin
  inherited;
  // Setup menu items
  self.menuEvalIndex := MENU_EVAL_INDEX;
  self.mainFormIcon := TIcon.Create();
  self.mainFormIcon.Handle := LoadImage(Hinstance, 'IDB_DBGP_ICO', IMAGE_ICON, 0, 0, (LR_DEFAULTSIZE or LR_LOADTRANSPARENT));
  self.UNCWarnings:= 0;

  // #112 = F1... pojma nimam od kje...
  self.PluginName := 'DBGp';
  self.AddFuncItem('&Debugger', _FuncDebugger);
  self.AddFuncItem('-', nil);

  sk := MakeShortcutKey(false, false, false, #118); // F7
  self.AddFuncItem('Step &Into', _FuncStepInto, sk);

  sk := MakeShortcutKey(false, false, false, #119); // F8
  self.AddFuncItem('Step &Over', _FuncStepOver, sk);

  sk := MakeShortcutKey(false, false, true, #119); // Shift+F8
  self.AddFuncItem('Step O&ut', _FuncStepOut, sk);

  self.AddFuncItem('Run &to', _FuncRunTo);

  sk := MakeShortcutKey(false, false, true, #120); // F9
  self.AddFuncItem('&Run', _FuncRun, sk);

  self.AddFuncItem('&Stop', _FuncStop);

  Assert(self.menuEvalIndex = self.AddFuncItem('-', nil));

  sk := MakeShortcutKey(true, false, false, #118); // Ctrl+F7
  self.AddFuncItem('&Eval', _FuncEval, sk);

  sk := MakeShortcutKey(true, false, false, #120); // Ctrl+F9
  self.AddFuncItem('Toggle &Breakpoint', _FuncBreakpoint, sk);

  // add stack and context items...

  self.AddFuncItem('-', nil);
  self.AddFuncItem('&Local Context', _FuncLocalContext);
  self.AddFuncItem('&Global Context', _FuncGlobalContext);
  self.AddFuncItem('Stac&k', _FuncStack);
  self.AddFuncItem('Break&points', _FuncBreakpoints);
  self.AddFuncItem('&Watches', _FuncWatches);
  self.AddFuncItem('-', nil);
  self.AddFuncItem('&Config', _FuncConfig);
  self.AddFuncItem('-', nil);
  self.AddFuncItem('&About...', _FuncAbout);
end;


destructor TDbgpNppPlugin.Destroy;
begin
  if (Assigned(self.MainForm)) then self.MainForm.Close;
  if (Assigned(self.mainFormIcon)) then FreeAndNil(self.mainFormIcon);
  inherited;
end;

{ hook }
procedure _FuncDebugger; cdecl;
begin
  Npp.FuncDebugger;
end;
procedure _FuncConfig; cdecl;
begin
  Npp.FuncConfig;
end;
procedure _FuncStepInto; cdecl;
begin
  Npp.FuncStepInto;
end;
procedure _FuncStepOver; cdecl;
begin
  Npp.FuncStepOver;
end;
procedure _FuncStepOut; cdecl;
begin
  Npp.FuncStepOut;
end;
procedure _FuncRunTo; cdecl;
begin
  Npp.FuncRunTo;
end;
procedure _FuncRun; cdecl;
begin
  Npp.FuncRun;
end;
procedure _FuncStop; cdecl;
begin
  Npp.FuncStop;
end;
procedure _FuncEval; cdecl;
begin
  Npp.FuncEval;
end;
procedure _FuncAbout; cdecl;
begin
  Npp.FuncAbout;
end;
procedure _FuncBreakpoint; cdecl;
begin
  Npp.FuncBreakpoint;
end;
procedure _FuncLocalContext; cdecl;
begin
  Npp.FuncLocalContext;
end;
procedure _FuncGlobalContext; cdecl;
begin
  Npp.FuncGlobalContext;
end;
procedure _FuncBreakpoints;
begin
  Npp.FuncBreakpoints;
end;
procedure _FuncStack;
begin
  Npp.FuncStack;
end;
procedure _FuncWatches;
begin
  Npp.FuncWatches;
end;

procedure TDbgpNppPlugin.FuncDebugger;
begin
  if (not IsCompatible) then Exit;

  self.ReadMaps(self.config.maps);
  // do some menu related stuff - njah...
  self.ChangeMenu(dmsDisconnected);
  if (Assigned(self.MainForm)) then
  begin
    self.MainForm.Show;
    exit;
  end;
  self.MainForm := TNppDockingForm1.Create(self);
  //self.MainForm.DlgId := self.FuncArray[0].CmdID;
  self.MainForm.DlgId := 0;
  self.MainForm.Icon := self.mainFormIcon;
  //self.MainForm.Show;
  self.MainForm.RegisterDockingForm(DWS_DF_CONT_BOTTOM);
  self.MainForm.Visible := true;
  if Assigned(self.MainForm.ServerSocket1) then
    self.MainForm.ServerSocket1.Port := self.config.listen_port;
  if (not self.config.start_closed) then self.MainForm.BitBtnCloseClick(nil); // activate socket
end;

procedure TDbgpNppPlugin.FuncAbout;
begin
  self.AboutForm := TAboutForm1.Create(self);
  self.AboutForm.DlgId := self.FuncArray[11].CmdID;
  self.AboutForm.Hide;
  self.AboutForm.ShowModal;
end;

procedure TDbgpNppPlugin.FuncConfig;
var
  r: TModalResult;
  port: integer;
begin
  if (not IsCompatible) then Exit;

  self.ReadMaps(self.config.maps);
  port := self.config.listen_port;
  self.ConfigForm := TConfigForm1.Create(self);
  //self.ConfigForm.DlgId := self.FuncArray[9].CmdID;
  self.ConfigForm.Hide;
  r := self.ConfigForm.ShowModal;
  self.ConfigForm := nil;
  if (r = mrOK) and (self.config.listen_port <> port) and Assigned(self.MainForm) then
  begin
    Warn('Restart the debugger to use port '+IntToStr(self.config.listen_port),
      'Port number changed', MB_ICONINFORMATION);
  end;
end;

procedure TDbgpNppPlugin.FuncEval;
begin
  // show eval dlg...
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.DoEval;
end;

procedure TDbgpNppPlugin.FuncBreakpoint;
begin
  if (IsCompatible and Assigned(self.MainForm)) and (self.MainForm.BitBtnBreakpoint.Enabled) then self.MainForm.BitBtnBreakpointClick(nil);
end;

procedure TDbgpNppPlugin.FuncRunTo;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.BitBtnRunToClick(nil);
end;

procedure TDbgpNppPlugin.FuncRun;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.DoResume(Run);
end;

procedure TDbgpNppPlugin.FuncStop;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.BitBtnStopClick(nil);
end;

procedure TDbgpNppPlugin.FuncStepInto;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.DoResume(StepInto);
end;

procedure TDbgpNppPlugin.FuncStepOut;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.DoResume(StepOut);
end;

procedure TDbgpNppPlugin.FuncStepOver;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.DoResume(StepOver);
end;

procedure TDbgpNppPlugin.FuncLocalContext;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.Open(dctLocalContect, true);
end;

procedure TDbgpNppPlugin.FuncGlobalContext;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.Open(dctGlobalContext, true);
end;

procedure TDbgpNppPlugin.FuncBreakpoints;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.Open(dctBreakpoints, true);
end;

procedure TDbgpNppPlugin.FuncStack;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.Open(dctStack, true);
end;

procedure TDbgpNppPlugin.FuncWatches;
begin
  if (IsCompatible and Assigned(self.MainForm)) then self.MainForm.Open(dctWatches, true);
end;

procedure TDbgpNppPlugin.InitMarkers;
var
  test: array [0..18] of AnsiString;
  r: integer;
begin
  r := SendMessage(CurrentScintilla, SCI_GETMARGINMASKN, 1, 0);
  r := r or (1 shl MARKER_ARROW) or (1 shl MARKER_BREAK);
  SendMessage(CurrentScintilla, SCI_SETMARGINMASKN, 1{_SC_MARGE_SYBOLE}, r);
  SendMessage(CurrentScintilla, SCI_MARKERSETALPHA, MARKER_ARROW, 90);

  SendMessage(CurrentScintilla, SCI_MARKERDEFINE,  MARKER_ARROW, SC_MARK_SHORTARROW{SC_MARK_ARROW});
  SendMessage(CurrentScintilla, SCI_MARKERSETFORE, MARKER_ARROW, $000000);
  SendMessage(CurrentScintilla, SCI_MARKERSETBACK, MARKER_ARROW, $00ff00);

  test[0]  := '14 14 3 1';
  test[1]  := ' 	c #FFFFFF';
  test[2]  := '.	c #000000';
  test[3]  := 'x	c #FF0000';
  test[4]  := '              ';
  test[5]  := '              ';
  test[6]  := '    ......    ';
  test[7]  := '   .xxxxxx.   ';
  test[8]  := '  .xxxxxxxx.  ';
  test[9]  := '  .xxxxxxxx.  ';
  test[10] := '  .xxxxxxxx.  ';
  test[11] := '  .xxxxxxxx.  ';
  test[12] := '  .xxxxxxxx.  ';
  test[13] := '  .xxxxxxxx.  ';
  test[14] := '   .xxxxxx.   ';
  test[15] := '    ......    ';
  test[16] := '              ';
  test[17] := '              ';

  SendMessage(CurrentScintilla, SCI_MARKERDEFINEPIXMAP,  MARKER_BREAK, LPARAM(@test));

  self.ChangeMenu(dmsOff);
end;

procedure TDbgpNppPlugin.ReadMaps(var maps: TMaps);
var
  path: String;
  ini: TIniFile;
  xmaps: TStringList;
  i: integer;
begin
  path := self.GetPluginsConfigDir;
  path := path + '\dbgp.ini';

  ini := TIniFile.Create(path);
  xmaps := TStringList.Create();
  ini.ReadSection('Mapping',xmaps);

  SetLength(maps, xmaps.Count);
  for i:=0 to xmaps.Count-1 do
  begin
    maps[i] := TStringList.Create;
    maps[i].Delimiter := ';';
    maps[i].DelimitedText := ini.ReadString('Mapping',xmaps[i],';;');
  end;

  self.config.refresh_local := ( ini.ReadString('Misc','refresh_local','0') = '1' );
  self.config.refresh_global := ( ini.ReadString('Misc','refresh_global','0') = '1' );
  self.config.use_source := ( ini.ReadString('Misc','use_source','0') = '1' );
  self.config.start_closed := ( ini.ReadString('Misc','start_closed','0') = '1' );
  self.config.break_first_line := ( ini.ReadString('Misc','break_first_line','0') = '1' );
  self.config.listen_port := ini.ReadInteger('Misc','listen_port',9003);
  self.config.max_depth := ini.ReadInteger('Features','max_depth',3);
  self.config.max_children := ini.ReadInteger('Features','max_children',15);
  self.config.local_setup := ( ini.ReadString('Misc','local_setup','1') = '1' );
  self.config.max_data := ini.ReadInteger('Features','max_data',512);

  ini.Free;
  xmaps.Free;
end;

procedure TDbgpNppPlugin.WriteMaps(conf:TDbgpNppPluginConfig);
var
  path: string;
  ini: TIniFile;
  xmaps: TStringList;
  i: integer;
begin
  path := self.GetPluginsConfigDir;
  if (not DirectoryExists(path)) then
  begin
    ForceDirectories(path);
  end;
  path := path + '\dbgp.ini';

  ini := TIniFile.Create(path);

  xmaps := TStringList.Create();
  ini.ReadSection('Mapping',xmaps);

  for i:=0 to xmaps.Count-1 do
  begin
    ini.DeleteKey('Mapping',xmaps[i]);
  end;
  xmaps.Free;

  for i:=0 to Length(conf.maps)-1 do
  begin
    if (conf.maps[i][0] = '') and (conf.maps[i][1] = '') and (conf.maps[i][2] = '') and (conf.maps[i][3] = '') then continue;
    conf.maps[i].Delimiter := ';';
    ini.WriteString('Mapping','Map'+IntToStr(i),conf.maps[i].DelimitedText);
  end;

  SetLength(TrueBoolStrs, 1);
  SetLength(FalseBoolStrs, 1);
  TrueBoolStrs[0] := '1';
  FalseBoolStrs[0] := '0';

  ini.WriteString('Misc','refresh_local',BoolToStr(conf.refresh_local, true));
  ini.WriteString('Misc','refresh_global',BoolToStr(conf.refresh_global, true));
  ini.WriteString('Misc','use_source',BoolToStr(conf.use_source, true));
  ini.WriteString('Misc','start_closed',BoolToStr(conf.start_closed, true));
  ini.WriteString('Misc','break_first_line',BoolToStr(conf.break_first_line, true));
  ini.WriteString('Misc','local_setup',BoolToStr(conf.local_setup, true));
  ini.WriteInteger('Misc','listen_port',conf.listen_port);

  ini.WriteInteger('Features','max_depth',conf.max_depth);
  ini.WriteInteger('Features','max_children',conf.max_children);
  ini.WriteInteger('Features','max_data',conf.max_data);

  ini.Free;

  // reread config
  self.ReadMaps(self.config.maps);
end;

// Test, za prikazovanje menujev
procedure TDbgpNppPlugin.GrayFuncItem(i: integer);
var
  hm: HMENU;
begin
  hm := GetMenu(self.NppData.NppHandle);
  EnableMenuItem(hm, self.FuncArray[i].CmdID, MF_BYCOMMAND or MF_DISABLED or MF_GRAYED);
end;

procedure TDbgpNppPlugin.EnableFuncItem(i: integer);
var
  hm: HMENU;
begin
  hm := GetMenu(self.NppData.NppHandle);
  EnableMenuItem(hm, self.FuncArray[i].CmdID, MF_BYCOMMAND or MF_ENABLED);
end;

procedure TDbgpNppPlugin.ChangeMenu(state: TDbgpMenuState);
var
  i: integer;
begin

  if (state = dmsOff) then
  begin
    for i:=2 to 7 do self.GrayFuncItem(i);
    for i:=self.menuEvalIndex to self.menuEvalIndex+1 do self.GrayFuncItem(i);
    for i:=self.menuEvalIndex+3 to self.menuEvalIndex+7 do self.GrayFuncItem(i);
  end;
  if (state = dmsConnected) then
  begin
    for i:=2 to 7 do self.EnableFuncItem(i);
    for i:=self.menuEvalIndex to self.menuEvalIndex+1 do self.EnableFuncItem(i);
    for i:=self.menuEvalIndex+3 to self.menuEvalIndex+7 do self.EnableFuncItem(i);
  end;
  if (state = dmsDisconnected) then
  begin
    for i:=2 to 7 do self.GrayFuncItem(i);
    for i:=self.menuEvalIndex to self.menuEvalIndex+1 do self.EnableFuncItem(i);
    for i:=self.menuEvalIndex+3 to self.menuEvalIndex+7 do self.EnableFuncItem(i);
  end;

end;

function TDbgpNppPlugin.Warn(Msg: String; Caption: String; Icon: Cardinal): Integer;
begin
  Result := MessageBoxW(Npp.NppData.NppHandle, PWChar(Msg), PWChar(Caption), Icon or MB_OK);
end;

procedure TDbgpNppPlugin.WarnUser;
const
  Msg: string = 'This version of DBGp requires Notepad++ 8.3 or newer.'#13#10
                 + 'Plugin commands have been disabled to prevent a crash.';
begin
  try
    if not self.IsCompatible then begin
      MessageBox(Npp.NppData.NppHandle, PChar(Msg), PChar('DBGp plugin (64-bit)'), MB_ICONWARNING);
    end;
  except
    ShowException(ExceptObject, ExceptAddr);
  end;
end;

end.
